### PR TITLE
Add depends_on support for steps

### LINKED ISSUE #1860: drop step.group in favour of depends_on to create a DAG
Support parallelism inside a pipeline via a [DAG](https://en.wikipedia.org/wiki/Directed_acyclic_graph). This requires a way to define step dependencies which can then allow the runner to build said DAG. Edit: I see [`group`](https://woodpecker-ci.org/docs/usage/pipeline-syntax#group---parallel-execution) provides something similar for static fan-out, but certainly not as flexible as defining dependencies directly.

_Originally posted by @silverwind in https://github.com/woodpecker-ci/woodpecker/issues/567#issuecomment-1282954563_

### PR DESCRIPTION
**Pull request by anbraten**
closes #1860

# TODO
- [x] add tests
- [x] add `depends_on` to schema
- [x] add `depends_on` to docs
- [x] fail if `depends_on` ref does not exist
- [x] add deprecation warning to linter for `groups`
- [x] prevent cyclic `depends_on`

### CODE REVIEW

**Comment by anbraten** [comment]
@6543 I've tried your approach using the `dagger` lib from #2184, but it was way more complex than implementing a simple topological search directly.

File: cli/lint/lint.go [hunk 1]
```diff
@@ -102,9 +102,9 @@ func lintFile(_ *cli.Context, file string) error {
 	// TODO: lint multiple files at once to allow checks for sth like "depends_on" to work
 	err = linter.New(linter.WithTrusted(true)).Lint([]*linter.WorkflowConfig{config})
 	if err != nil {
-		fmt.Printf("🔥 %s has errors:\n", output.String(config.File).Underline())
+		fmt.Printf("🔥 %s has warning / errors:\n", output.String(config.File).Underline())
```
Comment (anbraten) [inline cli/lint/lint.go:105]:
```suggestion
		fmt.Printf("🔥 %s has warnings / errors:\n", output.String(config.File).Underline())
```

File: docs/docs/20-usage/20-workflow-syntax.md [hunk 2]
```diff
@@ -444,33 +444,28 @@ when:
   - evaluate: 'SKIP != "true"'
 ```
 
-### `group` - Parallel execution
+### `depends_on`
 
-Woodpecker supports parallel step execution for same-machine fan-in and fan-out. Parallel steps are configured using the `group` attribute. This instructs the agent to execute the named group in parallel.
-
-Example parallel configuration:
+Normally steps of a workflow are executed serially in the order in which they are defined. As soon as you set `depends_on` for a step a [directed acyclic graph](https://en.wikipedia.org/wiki/Directed_acyclic_graph), all steps of the workflow will be executed in parallel besides the steps that have a dependency set to another step using `depends_on`:
```
Comment (anbraten) [inline docs/docs/20-usage/20-workflow-syntax.md:449]:
```suggestion
Normally steps of a workflow are executed serially in the order in which they are defined. As soon as you set `depends_on` for a step a [directed acyclic graph](https://en.wikipedia.org/wiki/Directed_acyclic_graph) will be used and all steps of the workflow will be executed in parallel besides the steps that have a dependency set to another step using `depends_on`:
```

File: pipeline/frontend/yaml/compiler/compiler.go [hunk 3]
```diff
@@ -328,3 +355,75 @@ func (c *Compiler) setupCacheRebuild(conf *yaml_types.Workflow, ir *backend_type
...
+		if err := dfsVisit(steps, dep, visited, path); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func convertDAGToStages(steps map[string]*stepWithDependsOn) ([]*backend_types.Stage, error) {
+	// Initialize the levels map
```
Comment (anbraten) [inline pipeline/frontend/yaml/compiler/compiler.go:377]:
```suggestion
```

File: pipeline/frontend/yaml/types/container.go [hunk 4]
```diff
@@ -48,6 +48,7 @@ type (
 		Volumes        Volumes            `yaml:"volumes,omitempty"`
 		When           constraint.When    `yaml:"when,omitempty"`
 		Ports          []base.StringOrInt `yaml:"ports,omitempty"`
+		DependsOn      []string           `yaml:"depends_on,omitempty"`
```
[the remark this answers is not in the document]
Comment (anbraten) [reply pipeline/frontend/yaml/types/container.go]: adjusted

**Comment by 6543** [comment]
I'l try to review it today :)

**Comment by lafriks** [comment]
Maybe this library could be useful for this - github.com/dominikbraun/graph?

**Comment by 6543** [comment]
> I'l try to review it today :)

well worked out quite nice - lets see if that today is today :/

File: cli/lint/lint.go [hunk 5]
```diff
@@ -101,9 +101,9 @@ func lintFile(_ *cli.Context, file string) error {
 	// TODO: lint multiple files at once to allow checks for sth like "depends_on" to work
 	err = linter.New(linter.WithTrusted(true)).Lint([]*linter.WorkflowConfig{config})
 	if err != nil {
-		fmt.Printf("🔥 %s has errors:\n", output.String(config.File).Underline())
+		fmt.Printf("🔥 %s has warnings / errors:\n", output.String(config.File).Underline())
 
-		hasErrors := true
+		hasErrors := false
```
Comment (6543) [inline cli/lint/lint.go:106]: unrelated fix - could move into own pull to get merged sooner

File: pipeline/backend/docker/docker.go [hunk 6]
```diff
@@ -318,10 +318,10 @@ func (e *docker) DestroyWorkflow(_ context.Context, conf *backend.Config, taskUU
 		for _, step := range stage.Steps {
 			containerName := toContainerName(step)
 			if err := e.client.ContainerKill(noContext, containerName, "9"); err != nil && !isErrContainerNotFoundOrNotRunning(err) {
-				log.Error().Err(err).Msgf("could not kill container '%s'", stage.Name)
+				log.Error().Err(err).Msgf("could not kill container '%s'", step.Name)
```
Comment (6543) [inline pipeline/backend/docker/docker.go:321]: nice catch but same as first coment

File: pipeline/frontend/yaml/compiler/compiler.go [hunk 7]
```diff
@@ -328,3 +355,74 @@ func (c *Compiler) setupCacheRebuild(conf *yaml_types.Workflow, ir *backend_type
...
+	for _, dep := range steps[name].dependsOn {
+		if err := dfsVisit(steps, dep, visited, path); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func convertDAGToStages(steps map[string]*stepWithDependsOn) ([]*backend_types.Stage, error) {
```
Comment (6543) [inline pipeline/frontend/yaml/compiler/compiler.go:376]: well I like to get rid of stages in the backend and only use a DAG independent of if we have it sequential or via group or via dag.

for now this should do but it wont do a full parallelism :/

Comment (anbraten) [reply pipeline/frontend/yaml/compiler/compiler.go:376]: Considered and refactored it to that in the beginning as well, but I would actually stick to the "stages" representation as its pretty nice to check the dependencies this way.

Comment (6543) [reply pipeline/frontend/yaml/compiler/compiler.go:376]: but it it will make a dag into "group"'s again and we loose the main feature of the dag so then we could just stick with groups

Comment (6543) [reply pipeline/frontend/yaml/compiler/compiler.go:376]: Aniway that should not bloc p.k fronted changes as p. backend can be changed independent

Comment (anbraten) [reply pipeline/frontend/yaml/compiler/compiler.go:376]: Sure, we can adjust that later on. Anything else missing for an approval?

Comment (6543) [reply pipeline/frontend/yaml/compiler/compiler.go:376]: let me review it :)

File: cli/lint/lint.go (hunk 5 above)
Comment (6543) [reply cli/lint/lint.go:106]: https://github.com/woodpecker-ci/woodpecker/pull/2995

File: pipeline/backend/docker/docker.go (hunk 6 above)
Comment (6543) [reply pipeline/backend/docker/docker.go:321]: that nit could stay in ... but I also put it into https://github.com/woodpecker-ci/woodpecker/pull/2995 as it lowers the pull diff here

File: pipeline/frontend/yaml/compiler/compiler.go [hunk 8]
```diff
@@ -234,8 +239,11 @@ func (c *Compiler) Compile(conf *yaml_types.Workflow) (*backend_types.Config, er
 	}
 
 	// add pipeline steps. 1 pipeline step per stage, at the moment
+	stepStages := make([]*backend_types.Stage, 0)
```
Comment (6543) [inline pipeline/frontend/yaml/compiler/compiler.go:242]:
```suggestion
	stepStages := make([]*backend_types.Stage, 0, 1)
```

File: pipeline/frontend/yaml/compiler/compiler.go [hunk 9]
```diff
@@ -274,9 +286,24 @@ func (c *Compiler) Compile(conf *yaml_types.Workflow) (*backend_types.Config, er
...
+		steps[container.Name] = &stepWithDependsOn{
+			step:      step,
+			dependsOn: container.DependsOn,
+		}
 		stage.Steps = append(stage.Steps, step)
 	}
 
+	// dag is used if one or more steps have a depends_on
+	if useDag {
+		stages, err := convertDAGToStages(steps)
```
Comment (6543) [inline pipeline/frontend/yaml/compiler/compiler.go:298]: looks like compile test contains no  convertDAGToStages fails

File: pipeline/frontend/yaml/compiler/compiler.go [hunk 10]
```diff
@@ -328,3 +355,74 @@ func (c *Compiler) setupCacheRebuild(conf *yaml_types.Workflow, ir *backend_type
 
 	return nil
 }
+
+func dfsVisit(steps map[string]*stepWithDependsOn, name string, visited map[string]struct{}, path []string) error {
```
Comment (6543) [inline pipeline/frontend/yaml/compiler/compiler.go:359]: could we move this into own file?

e.g. dag.go ?

**Comment by 6543** [comment]
@anbraten addressed all my review things at https://github.com/anbraten/woodpecker/pull/5 and more

File: pipeline/frontend/yaml/compiler/compiler.go (hunk 10 above)
Comment (6543) [reply pipeline/frontend/yaml/compiler/compiler.go:359]: -> https://github.com/anbraten/woodpecker/pull/5

File: pipeline/frontend/yaml/compiler/compiler.go (hunk 9 above)
Comment (6543) [reply pipeline/frontend/yaml/compiler/compiler.go:298]: -> https://github.com/anbraten/woodpecker/pull/5

File: pipeline/frontend/yaml/compiler/compiler.go (hunk 8 above)
Comment (6543) [reply pipeline/frontend/yaml/compiler/compiler.go:242]: -> https://github.com/anbraten/woodpecker/pull/5

**Review by 6543 — APPROVED** [review]

**Review by 6543 — APPROVED** [review]
